README.md

```diff
@@ -1,68 +1,256 @@
+   (with an appended PR number) and the PR description as the commit
+   message. TODO: What happens to long paragraph lines often present in
+   the PR description? We should either reject PRs that have long lines
+   or, if it can be done safely, automatically wrap them using some
+   formatting library.
+3. Test the staging branch. That is, wait for GitHub to report CI test
+   results for the staging branch.
+4. Reset the PR target branch (usually "master") to point at the
+   now-tested commit on the staging branch.
+5. Mark the PR as merged.
+6. Close the PR and remove its tag. Failures at this step are ignored.
+
+XXX: The above is missing the tagging step, with a brief explanation of
+what that tag means and/or why it is needed.
+
+If a bot is killed while executing the above algorithm, it (XXX:
+document what happens. I suspect we redo some of the steps and resume
+execution of other steps, depending on the state, but it is not clear to
+me which steps the bot avoids repeating).
+
+While the bot is merging a pull request, it does not touch other PRs. If
+the bot receives a GitHub event while merging a pull request, then the
+bot re-examines all PRs _after_ it is done merging the current PR. This
+approach simplifies repository state (multiple PRs in various
+intermediate merging steps would easily confuse humans and may overwhelm
+CI) while still allowing the lowest-numbered PR to be merged first
+(among the then-ready PRs).
+
+
+## Error handling
+
+A bot that cannot parse its configuration quits.
+
+When encountering PR-unrelated problems (e.g., network I/O timeouts or
+internal GitHub server errors), the bot sleeps for 10 minutes
+(hard-coded) and then effectively restarts. Only the already parsed
+configuration is preserved across such restarts. XXX: Let's restart
+everything except configuration parsing in this case instead of assuming
+that the global state such as the HTTP server object is still in a good
+working order.
+
+If a PR processing step fails for PR-specific reasons (e.g., a CI test
+failure), then the bot moves on to the next pull request, marking the
+failed PR if/as needed (see below for PR labels).
+
+
+## Pull request labels
+
+The bot uses the following GitHub labels to mark the current pull
+request state:
+
+* S-merging: The PR is being processed by the bot. The bot eventually
+  replaces this label with one of the other labels below.
+* S-staging-checks-failed: Essentially duplicates GitHub "red x" mark for
+  the staging branch commit. The bot removes this label when it notices
+  that the failed checks are no longer fresh/applicable.
+* S-merge-ready: Duplicates GitHub "green check" mark for the staging
+  branch commit. The bot removes this label before it updates the PR
+  target branch. (TODO: Remove due to very short lifetime and being low
+  on information? All S-merged and S-merge-failed PRs are S-merge-ready
+  for a brief moment...)
```

Comment:
This label primary purpose is to mark our 'dry run 2' mode (when all staging checks have passed but we skipped the last step). Without dry run, it will disappear very quickly, as you noted.

Reply:
I fixed this in a way we discussed: mark only when 'dry run 2' mode is on.